lib/src/main/java/tech/relaycorp/relaydroid/messaging/ReceiveMessages.kt

```diff
@@ -1,27 +1,36 @@
 package tech.relaycorp.relaydroid.messaging
 
+import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.asFlow
 import kotlinx.coroutines.flow.flatMapLatest
-import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.mapNotNull
 import tech.relaycorp.poweb.PoWebClient
 import tech.relaycorp.relaydroid.Relaynet
 import tech.relaycorp.relaydroid.Storage
+import tech.relaycorp.relaydroid.common.Logging.logger
 import tech.relaycorp.relaynet.bindings.pdc.PDCClient
 import tech.relaycorp.relaynet.bindings.pdc.Signer
 import tech.relaycorp.relaynet.bindings.pdc.StreamingMode
-import tech.relaycorp.relaynet.messages.Parcel
+import tech.relaycorp.relaynet.ramf.RAMFException
+import java.util.logging.Level
 
 internal class ReceiveMessages(
     private val pdcClientBuilder: () -> PDCClient = { PoWebClient.initLocal(Relaynet.POWEB_PORT) }
 ) {
 
-    fun receive() =
+    fun receive(): Flow<IncomingMessage> =
         getNonceSigners()
             .flatMapLatest { nonceSigners ->
                 pdcClientBuilder().use {
                     it.collectParcels(nonceSigners, StreamingMode.CloseUponCompletion)
-                        .map { parcelCollection ->
-                            val parcel = Parcel.deserialize(parcelCollection.parcelSerialized)
+                        .mapNotNull { parcelCollection ->
+                            val parcel = try {
+                                parcelCollection.deserializeAndValidateParcel()
+                            } catch (exp: RAMFException) {
```

Comment:
You need to catch `RelaynetException`, or create a separate `catch` block for `InvalidMessageException`.

`RAMFException` is only for malformed parcels. `InvalidMessageException` is for well-formed yet invalid messages.